mesonbuild/wrap/wrap.py

```diff
@@ -136,47 +136,64 @@ class WrapException(MesonException):
 class WrapNotFoundException(WrapException):
     pass
 
+@dataclasses.dataclass(eq=False)
 class PackageDefinition:
-    def __init__(self, fname: str, subproject: str = ''):
-        self.filename = fname
-        self.subproject = SubProject(subproject)
-        self.type: T.Optional[str] = None
-        self.values: T.Dict[str, str] = {}
-        self.provided_deps: T.Dict[str, T.Optional[str]] = {}
-        self.provided_programs: T.List[str] = []
-        self.diff_files: T.List[Path] = []
-        self.basename = os.path.basename(fname)
-        self.has_wrap = self.basename.endswith('.wrap')
-        self.name = self.basename[:-5] if self.has_wrap else self.basename
-        # must be lowercase for consistency with dep=variable assignment
-        self.provided_deps[self.name.lower()] = None
-        # What the original file name was before redirection
-        self.original_filename = fname
-        self.redirected = False
-        if self.has_wrap:
-            self.parse_wrap()
-            with open(fname, 'r', encoding='utf-8') as file:
-                self.wrapfile_hash = hashlib.sha256(file.read().encode('utf-8')).hexdigest()
+    name: str
+    subprojects_dir: str
+    type: T.Optional[str] = None
+    values: T.Dict[str, str] = dataclasses.field(default_factory=dict)
+    provided_deps: T.Dict[str, T.Optional[str]] = dataclasses.field(default_factory=dict)
+    provided_programs: T.List[str] = dataclasses.field(default_factory=list)
+    diff_files: T.List[Path] = dataclasses.field(default_factory=list)
+    wrapfile_hash: T.Optional[str] = None
+    original_filename: T.Optional[str] = None
+    filedir: str = dataclasses.field(init=False)
+    directory: str = dataclasses.field(init=False)
```

Comment:
Converting all of these into dataclass init attributes feels like going in the exact oppposite direction of what we want.

Dataclasses exist largely to handle boilerplate logic when accepting values in arguments to `__init__()` and setting them to `self.*` with custom logic for sentinel "not-passed" that doesn't corrupt the attribute API, etc.

It's not really the right tool for when your entire class is just a bunch of properties set by complex logic after the fact.

Reply:
You can do that without dataclasses, surely?

Just add a staticmethod constructor. Those don't rely on dataclasses.

Reply:
What's wrong with using dataclass, it makes it cleaner than writing by hand
```
    def __init__(self, name: str, subprojects_dir: str, type_: T.Optional[str] = None, values: T.Optional[T.Dict[str, str]] = None):
        self.name = name
        ...
```

Reply:
Because the dataclass means the function is actually:

```
def __init__(self,  name: str, subprojects_dir: str, type: T.Optional[str] = None, values: T.Dict[str, str] = None, provided_deps: T.Dict[str, T.Optional[str]] = None, provided_programs: T.List[str] = None, diff_files: T.List[Path] = None,  wrapfile_hash: T.Optional[str] = None, original_filename: T.Optional[str] = None):
```

And it's... not actually clear we want that?

Reply:
That does not seems to be a big deal. If that was C++ I would just make that constructor private, but I don't think Python can. I like to have every public field in top like dataclass does, it makes usage of that class easier IMHO.

Reply:
With a custom `__ini__` you have to search for attributes set on self, and unless we're consistent with usage of underscore to mark private fields (we never are), it's difficult to know what can l should be accessed or not.

Reply:
There is absolutely NO connection between whether something is public/private and whether something is a dataclass.

Dataclasses support underscored private fields just fine:
```python
@dataclass
class Foo:
    _foo: int
    bar: int
```
```python
>>> c = Foo(1, 2)
>>> c.<TAB>
c.bar
```

No public access to _foo anywhere...

Reply:
Sure, but that still doesn't change that using dataclass is much cleaner.  But if you insist on writing less readable code, that's not a hill I'm ready to die on... as you wish I don't care.